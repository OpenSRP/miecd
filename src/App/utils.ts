import { DEFAULT_PLAN_ID, KEYCLOAK_API_BASE_URL, KEYCLOAK_USERS_PAGE_SIZE, OPENSRP_API_BASE_URL } from 'configs/env';
import { URL_LOCATION_UNIT } from '../constants';

const openSrpRestBase = OPENSRP_API_BASE_URL + '/rest/';

export const baseProps = {
    baseURL: openSrpRestBase,
    opensrpBaseURL: openSrpRestBase,
    keycloakBaseURL: KEYCLOAK_API_BASE_URL,
};

export const locationUnitProps = {
    ...baseProps,
    filterByParentId: false,
};

export const newLocationUnitProps = {
    ...locationUnitProps,
    successURLGenerator: () => URL_LOCATION_UNIT,
    cancelURLGenerator: () => URL_LOCATION_UNIT,
    hidden: ['serviceType', 'latitude', 'longitude'],
};

export const editLocationProps = {
    ...newLocationUnitProps,
    ...locationUnitProps,
};

export const usersListProps = {
    ...baseProps,
    usersPageSize: KEYCLOAK_USERS_PAGE_SIZE,
};

export const teamAssignmentProps = {
    ...baseProps,
    defaultPlanId: DEFAULT_PLAN_ID,
};

export const createEditUserProps = {
    ...baseProps,
    userFormHidden: [],
};
