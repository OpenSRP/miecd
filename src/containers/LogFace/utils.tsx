import { replace, split, trim } from 'lodash';
import React from 'react';

/** convert the smsData message field from prose to more easily readable
 * point format
 *
 * @param message - the message prose.
 */
export const parseMessage = (message: string) => {
    const propValues = split(message, '\n');
    const replacedEquals = propValues.map(trim).map((entry) => replace(entry, / =\s*/, ' : '));
    return (
        <ul>
            {replacedEquals.map((value, index) => {
                return <li key={`${value}-${index}`}>{value}</li>;
            })}
        </ul>
    );
};
