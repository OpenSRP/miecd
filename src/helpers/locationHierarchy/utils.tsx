import { ParsedHierarchyNode, RawHierarchyNode, RawHierarchyNodeMap, RawOpenSRPHierarchy, TreeNode } from './types';
import { cloneDeep } from 'lodash';
import cycle from 'cycle';
import TreeModel from 'tree-model';

/** Parse the raw child hierarchy node map
 *
 * @param {RawHierarchyNodeMap} rawNodeMap - Object of raw hierarchy nodes
 * @param {string} parent - node parent id
 * @returns {Array<ParsedHierarchyNode>} Array of Parsed hierarchy nodes
 */
const parseChildren = (rawNodeMap: RawHierarchyNodeMap, parent: string) => {
    // skip key when destructuring
    const rawHierarchyNode: RawHierarchyNode[] = Object.entries(rawNodeMap).map(([, value]) => value);
    return rawHierarchyNode.map((child) => {
        const parsedNode: ParsedHierarchyNode = {
            ...child,
            title: child.label,
            key: `${child.id}-${parent}`,
            children: child.children ? parseChildren(child.children, parent) : [],
        };
        return parsedNode;
    });
};

/** parses the raw opensrp hierarchy to a hierarchy that we can quickly build
 * our tree model from.
 *
 * @param {RawOpenSRPHierarchy} raw - the response we get from opensrp
 * @returns {ParsedHierarchyNode} - returns Parent node with its children
 */
const parseHierarchy = (raw: RawOpenSRPHierarchy) => {
    // clone the locationTree, we are going to be mutating a copy
    const rawClone: RawOpenSRPHierarchy = cloneDeep(raw);

    // !IMPORTANT ASSUMPTION : locationsTreeClone has a single object under map, i.e there is only one root jurisdiction
    const { map } = rawClone.locationsHierarchy;
    // !IMPORTANT ASSUMPTION : locationsTreeClone has a single object under map, i.e there is only one root jurisdiction
    // skip key when destructuring
    const rawNode: RawHierarchyNode = Object.entries(map).map(([, value]) => value)[0];
    const parsedNode: ParsedHierarchyNode = {
        ...rawNode,
        title: rawNode.label,
        key: rawNode.id,
        children: rawNode.children ? parseChildren(rawNode.children, rawNode.id) : [],
    };

    return parsedNode;
};

/** takes the raw opensrp hierarchy response and creates a tree model structure
 *
 * @param {RawOpenSRPHierarchy} apiResponse - the response we get from opensrp
 * @returns {TreeNode} - returns root node
 */
export const generateJurisdictionTree = (apiResponse: RawOpenSRPHierarchy): TreeNode => {
    const tree = new TreeModel();
    const hierarchy = parseHierarchy(apiResponse);
    const root = tree.parse<ParsedHierarchyNode>(hierarchy);
    return root;
};

/**
 * serialize tree due to circular dependencies
 *
 * @param trees - trees to be serialized
 */
export const serializeTree = (trees: (TreeNode | undefined)[]) => {
    return JSON.stringify(trees.map((tree) => JSON.stringify(cycle.decycle(tree))));
};
