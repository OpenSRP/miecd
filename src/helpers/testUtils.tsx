import { mount, MountRendererProps } from 'enzyme';
import React from 'react';
import { I18nextProvider } from 'react-i18next';
import i18n from '../mls';
import { act } from 'react-dom/test-utils';

export const mountWithTranslations: typeof mount = (component: React.ReactNode, options?: MountRendererProps) => {
    return mount(<I18nextProvider i18n={i18n}>{component}</I18nextProvider>, options);
};

/**
 * Utility to await promises to resolve in tests
 * @param ms time in ms
 */
export async function waitForPromises() {
    await act(async () => {
        await new Promise((res) => setImmediate(res));
    });
}
