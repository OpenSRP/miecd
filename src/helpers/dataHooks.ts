import {
    OPENSRP_SECURITY_AUTHENTICATE,
    LOCATION_FILTER_PARAM,
    RISK_CATEGORY_FILTER_PARAM,
    SMS_TYPE_FILTER_PARAM,
    SEARCH_FILTER_PARAM,
    PREGNANCY_DETECTION,
    UPDATE_PREGNANCY_DETECTION,
    NUTRITION_REGISTRATION,
    UPDATE_NUTRITION_REGISTRATION,
    DEATH_REPORT,
    UPDATE_DEATH_REPORT,
} from '../constants';
import { getQueryParams } from 'containers/LogFace/utils';
import { TFunction } from 'i18next';
import { useQuery } from 'react-query';
import { useSelector } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { OpenSRPService } from 'services/opensrp';
import { getSmsDataByFilters } from 'store/ducks/sms_events';
import { TreeNode } from './locationHierarchy/types';
import { generateJurisdictionTree } from './locationHierarchy/utils';
import { toastToError, getRiskCatFilter } from './utils';
import { LogFaceModules } from 'configs/settings';

const selectSmsData = getSmsDataByFilters();

/** hook to get this users assignment details, hierarchy, team assigned and location the team is assigned
 * @param t - the translator function
 */
export const useUserAssignment = (t: TFunction) => {
    return useQuery(
        'userAssignment',
        () => {
            const opensrpService = new OpenSRPService(OPENSRP_SECURITY_AUTHENTICATE);
            return opensrpService.read('');
        },
        {
            select: (response) => {
                return {
                    userHierarchy: generateJurisdictionTree(response.locations),
                    user: response.user,
                    team: response.team.team,
                    location: response.team.team.location,
                };
            },
            onError: (err: Error) => toastToError(err?.message ?? t('An error occurred')),
        },
    );
};

/** redux hok to select filtered sms data based on provided filters
 * @param location - route location details, to get search params
 * @param module - the module
 * @param userLocationTree - for filtering based on the user's assigned level
 */
export const useGetLogFaceFilteredSms = (
    location: RouteComponentProps['location'],
    module: LogFaceModules,
    userHierarchy?: TreeNode,
) => {
    return useSelector((state) => {
        const userLocationIdFilter = getQueryParams(location)[LOCATION_FILTER_PARAM] as string;
        const riskCategoryFilter = getQueryParams(location)[RISK_CATEGORY_FILTER_PARAM] as string;
        const smsTypeFilter = getQueryParams(location)[SMS_TYPE_FILTER_PARAM] as string;
        const searchFilter = getQueryParams(location)[SEARCH_FILTER_PARAM] as string;
        const matchesId = (node: TreeNode) => node.model.id === userLocationIdFilter;
        const userLocationNode = userHierarchy?.first(matchesId);
        let smsFilter = smsTypeFilter ? [smsTypeFilter] : undefined;
        switch (smsTypeFilter) {
            case PREGNANCY_DETECTION:
                smsFilter = [PREGNANCY_DETECTION, UPDATE_PREGNANCY_DETECTION];
                break;
            case NUTRITION_REGISTRATION:
                smsFilter = [NUTRITION_REGISTRATION, UPDATE_NUTRITION_REGISTRATION];
                break;
            case DEATH_REPORT:
                smsFilter = [DEATH_REPORT, UPDATE_DEATH_REPORT];
                break;
            default:
                break;
        }
        return selectSmsData(state, {
            locationNode: userLocationNode,
            riskCategory: getRiskCatFilter(module, riskCategoryFilter),
            smsTypes: smsFilter,
            searchFilter,
            module,
        });
    });
};
