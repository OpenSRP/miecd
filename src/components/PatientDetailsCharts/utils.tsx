import NoRecord from 'components/NoRecord';
import Ripple from 'components/page/Loading';
import { TFunction, useTranslation } from 'react-i18next';
import supersetFetch from 'services/superset';
import { getMonthNames } from '../../constants';
import { sortByEventDate } from '../../helpers/utils';
import { ChildChartData, MotherChartData } from 'store/ducks/chartData';
import superset, { SupersetFormData } from '@onaio/superset-connector';
import { ActionCreator } from 'redux';
import { GET_FORM_DATA_ROW_LIMIT } from 'configs/env';
import React from 'react';

/** factory for form data filters for chart pages */
export const chartFormDataFilters = (patientId: string) =>
    superset.getFormData(GET_FORM_DATA_ROW_LIMIT, [
        {
            comparator: patientId,
            operator: '==',
            subject: 'patient_id',
        },
    ]);

/** abstracts code that actually makes the superset Call since it is quite similar
 * @param supersetSlice - slice string
 * @param actionCreator - redux action creator
 * @param supersetService - the supersetService
 * @param supersetOptions - adhoc filters for superset call
 */
export async function chartSupersetCall<TAction, TResponse>(
    supersetSlice: string,
    actionCreator: ActionCreator<TAction>,
    supersetService: typeof supersetFetch = supersetFetch,
    supersetOptions: SupersetFormData | null = null,
) {
    const asyncOperation = supersetOptions
        ? supersetService(supersetSlice, supersetOptions)
        : supersetService(supersetSlice);
    return asyncOperation
        .then((result: TResponse[] | undefined) => {
            if (result) {
                actionCreator(result);
            }
        })
        .catch((error) => {
            throw error;
        });
}

/** props for BrokenLoading chart */
export interface BrokenLoadingChartPageProps {
    error?: Error;
    loading: boolean;
    broken: boolean;
}

/** dry component for showing loader or broken page */
export const BrokenLoadingChartPage = ({ error, loading, broken }: BrokenLoadingChartPageProps) => {
    const { t } = useTranslation();
    if (loading) {
        return <Ripple></Ripple>;
    }

    if (broken) {
        return <NoRecord message={t(`Chart could not load, an error occurred. ${error?.message ?? ''}`)} />;
    }
    return null;
};

/** create a chart data series containing the weight and height data series from smsEvents */
export const getWeightHeightDataSeries = (smsEvents: ChildChartData[], t: TFunction) => {
    const categories: string[] = [];
    const dataSeries: { data: (number | undefined)[]; name: string }[] = [
        { data: [], name: 'weight' },
        { data: [], name: 'height' },
    ];
    // make sure events are sorted from oldest to latest
    const sortedSmsEvents = sortByEventDate(smsEvents).reverse();

    for (const data of sortedSmsEvents) {
        const calendarCategoryObj = {
            month: new Date(data.event_date).getMonth(),
            year: new Date(data.event_date).getFullYear(),
        };
        if (data.weight && data.height) {
            categories.push(`${getMonthNames(t)[calendarCategoryObj.month]}/${calendarCategoryObj.year}`);
            dataSeries[0].data.push(data.weight);
            dataSeries[1].data.push(data.height);
        }
    }
    return {
        categories,
        dataSeries,
    };
};

/** create a chart data series containing the weight data series from smsEvents */
export const getWeightDataSeries = (smsEvents: MotherChartData[], t: TFunction) => {
    const categories: string[] = [];
    const dataSeries: { data: (number | undefined)[]; name: string }[] = [{ data: [], name: 'weight' }];
    // make sure events are sorted from oldest to latest
    const sortedSmsEvents = sortByEventDate(smsEvents).reverse();

    for (const data of sortedSmsEvents) {
        const calendarCategoryObj = {
            month: new Date(data.event_date).getMonth(),
            year: new Date(data.event_date).getFullYear(),
        };
        if (data.weight) {
            categories.push(`${getMonthNames(t)[calendarCategoryObj.month]}/${calendarCategoryObj.year}`);
            dataSeries[0].data.push(data.weight);
        }
    }
    return {
        categories,
        dataSeries,
    };
};

/** create a data series containing the diastolic and systolic data series from smsEvents */
export const getBloodPSeriesForChart = (smsEvents: MotherChartData[], t: TFunction) => {
    const categories: string[] = [];
    const dataSeries: { data: (number | undefined)[]; name: string }[] = [
        { data: [], name: 'systolic' },
        { data: [], name: 'diastolic' },
    ];
    // make sure events are sorted from oldest to latest
    const sortedSmsEvents = sortByEventDate(smsEvents).reverse();

    const regex = /\d{2,3}\S{0,3}\d{2,3}/;
    const singleBpValueRegex = /\d{2,3}/g;

    for (const data of sortedSmsEvents) {
        const calendarCategoryObj = {
            month: new Date(data.event_date).getMonth(),
            year: new Date(data.event_date).getFullYear(),
        };
        const hasValidBp = regex.test(data.bp);
        if (!hasValidBp) {
            continue;
        }
        categories.push(`${getMonthNames(t)[calendarCategoryObj.month]}/${calendarCategoryObj.year}`);
        const matchedValues = [...Array.from(data.bp.matchAll(singleBpValueRegex))];

        dataSeries[0].data.push(Number(matchedValues[0]));
        dataSeries[1].data.push(Number(matchedValues[1]));
    }
    return {
        categories,
        dataSeries,
    };
};
