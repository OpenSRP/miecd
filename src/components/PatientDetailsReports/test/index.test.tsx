import { PatientDetailsReport, chunkByGravida, pregnancyOptionsFilter } from '..';
import { mountWithTranslations } from '../../../helpers/testUtils';
import { pregnancyReport1, pregnancyReports } from './fixtures';
import { Dictionary } from '@onaio/utils';
import React from 'react';
import { PregnancyReportFixture } from 'store/ducks/tests/fixtures';
import toJson from 'enzyme-to-json';
import { NutritionLogFaceSms } from 'store/ducks/sms_events';

jest.mock('react-select', () => ({ options, onChange }: Dictionary) => {
    function handleChange(event: Dictionary) {
        const option = options.find((option: Dictionary) => option.value === event.target?.value);
        onChange(option);
    }

    return (
        <select data-testid="select" onChange={handleChange}>
            {options.map(({ label, value }: Dictionary) => (
                <option key={value} value={value}>
                    {label}
                </option>
            ))}
        </select>
    );
});

describe('PatientDetailsReport', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    it('works okay', () => {
        const patientsReports = PregnancyReportFixture.filter(
            (sms) => sms.patient_id == '10063H',
        ) as NutritionLogFaceSms[];
        const props = {
            patientsReports,
            isChild: false,
        };
        const wrapper = mountWithTranslations(<PatientDetailsReport {...props} />);
        wrapper.find('table tr').forEach((tr) => {
            expect(tr.text()).toMatchSnapshot('initial current pregnancies');
        });
    });
    it('can switch pregnancies', () => {
        const patientsReports = PregnancyReportFixture.filter(
            (sms) => sms.patient_id == '10063H' || sms.patient_id === '101AVR',
        ) as NutritionLogFaceSms[];
        const props = {
            patientsReports,
            isChild: false,
        };
        const wrapper = mountWithTranslations(<PatientDetailsReport {...props} />);
        wrapper.find('table tr').forEach((tr) => {
            expect(tr.text()).toMatchSnapshot('initial current pregnancies');
        });
        // switch pregnancies
        expect(toJson(wrapper.find('select'))).toMatchSnapshot('select snapshot');
        wrapper.find('select').simulate('change', { target: { value: 1618905587899, name: 'pregnancy 1' } });
        wrapper.update();

        wrapper.find('table tr').forEach((tr) => {
            expect(tr.text()).toMatchSnapshot('after filter pregnancies');
        });
    });
});

describe('PatientDetailsReport utils', () => {
    const sampleT = (t: string) => t;

    it('able to group sms by pregnancy_id correctly', () => {
        const grouped = chunkByGravida(pregnancyReports);
        const expected = {
            [pregnancyReport1.pregnancy_id]: pregnancyReports,
        };
        expect(grouped).toEqual(expected);
    });
    it('pregnancyOptionsFilter', () => {
        const chunkedSms = chunkByGravida(pregnancyReports);
        const response = pregnancyOptionsFilter(chunkedSms, sampleT);
        expect(response).toEqual([{ label: 'Current pregnancy', value: '1618905587899' }]);
    });
});
