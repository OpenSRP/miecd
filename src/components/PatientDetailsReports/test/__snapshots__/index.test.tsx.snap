// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PatientDetailsReport can switch pregnancies: after filter pregnancies 1`] = `"ReportDateReporterMessage"`;

exports[`PatientDetailsReport can switch pregnancies: initial current pregnancies 1`] = `"ReportDateReporterMessage"`;

exports[`PatientDetailsReport can switch pregnancies: initial current pregnancies 2`] = `
"Birth Report2021-04-23Husna HarizChild DOB = 23-04-2021 
Child Gender = NU 
Delivery Location =  
Newborn Symptoms = Lack of movement 
Child Weight = 4.9 
Child Length = 50 
Child Risk =  
Breastfeeding = No breastfeeding within 1 hour
Mother Symptoms = Bu00ecnh thu01b0u1eddng"
`;

exports[`PatientDetailsReport can switch pregnancies: initial current pregnancies 3`] = `
"Birth Report2021-04-23Husna HarizChild DOB = 23-04-2021 
Child Gender = NU 
Delivery Location =  
Newborn Symptoms = Lack of movement 
Child Weight = 4.9 
Child Length = 50 
Child Risk =  
Breastfeeding = No breastfeeding within 1 hour
Mother Symptoms = Bu00ecnh thu01b0u1eddng"
`;

exports[`PatientDetailsReport can switch pregnancies: initial current pregnancies 4`] = `
"Pregnancy Detection2021-04-23Husna HarizName = BENT SENT  ALEXA
Phone Number = 0912782110 
Health Insurance = KCO"
`;

exports[`PatientDetailsReport can switch pregnancies: select snapshot 1`] = `
<select
  data-testid="select"
  onChange={[Function]}
>
  <option
    key="1619157940156"
    value="1619157940156"
  >
    Current pregnancy
  </option>
  <option
    key="1618905587899"
    value="1618905587899"
  >
    Pregnancy 1
  </option>
</select>
`;

exports[`PatientDetailsReport works okay: initial current pregnancies 1`] = `"ReportDateReporterMessage"`;

exports[`PatientDetailsReport works okay: initial current pregnancies 2`] = `
"Birth Report2021-04-22Husna HarizChild DOB = 22-04-2021 
Child Gender = NU 
Delivery Location =  
Newborn Symptoms = Lack of movement 
Child Weight = 4.9 
Child Length = 50 
Child Risk = low 
Breastfeeding = No breastfeeding within 1 hour
Mother Symptoms = Bu00ecnh thu01b0u1eddng"
`;

exports[`PatientDetailsReport works okay: initial current pregnancies 3`] = `
"Delivery Planning2021-04-22Husna HarizPlanned Delivery Location = Nhu00e0 
Current Symptoms = Ho nhiu1ec1u hou1eb7c ho khan, HNgu00f4i ngu01b0u1ee3c"
`;

exports[`PatientDetailsReport works okay: initial current pregnancies 4`] = `
"Home ANC Visit2021-04-21Husna HarizBlood Pressure = 118/78 
Weight = 70 
Current Symptoms = Bu00ecnh thu01b0u1eddng 
Supplements = Micronutrients Supplement: No 
Risk = low 
Contact = 3"
`;

exports[`PatientDetailsReport works okay: initial current pregnancies 5`] = `
"ANC Visit2021-04-21Husna HarizBlood Pressure = 118/78
Weight = 60
Current Symptoms = Bu00ecnh thu01b0u1eddng
Supplements = Micronutrients Supplement: No
Risk = low
Contact = 2"
`;

exports[`PatientDetailsReport works okay: initial current pregnancies 6`] = `
"ANC Visit2021-04-21Husna HarizBlood Pressure = 118/78
Weight = 58
Current Symptoms = Bu00ecnh thu01b0u1eddng
Supplements = Bu1ed5 sung vi chu1ea5t: Cu00f3
Risk = low
Contact = 1"
`;

exports[`PatientDetailsReport works okay: initial current pregnancies 7`] = `
"Social Determinants2021-04-21Husna HarizPlanned Delivery Location = Bu1ec7nh viu1ec7n tu1ec9nh hou1eb7c Bu1ec7nh viu1ec7n Trung u01afu01a1ng 
Household = Is a poor household
Ethnicity = TEST1 TEST2 TEST3
Toilet = Has toilet
Handwashing = No Handwashing facilities with soap"
`;

exports[`PatientDetailsReport works okay: initial current pregnancies 8`] = `
"Social Determinants2021-04-21Husna HarizPlanned Delivery Location = Bu1ec7nh viu1ec7n tu1ec9nh hou1eb7c Bu1ec7nh viu1ec7n Trung u01afu01a1ng 
Household = Is a poor household
Ethnicity = TEST1 TEST2 TEST3
Toilet = Has toilet
Handwashing = No Handwashing facilities with soap"
`;

exports[`PatientDetailsReport works okay: initial current pregnancies 9`] = `
"Pregnancy Registration2021-04-21Husna HarizEDD = 08-08-2021
Gravidity = 2
Parity = 1
Weight = 50.4 
Height = 123
BMI = 33.31350386674598
Previous Gynaecology History = Bu00ecnh thu01b0u1eddng
Current Symptoms = Bu00ecnh thu01b0u1eddng, Ho nhiu1ec1u hou1eb7c ho khan"
`;

exports[`PatientDetailsReport works okay: initial current pregnancies 10`] = `
"Pregnancy Detection2021-04-20Husna HarizName = TAH PENNY  ALICE
Phone Number = 0912312000 
Health Insurance = 1000010101"
`;
